tower-ready-cache/src/service.rs

```diff
@@ -0,0 +1,380 @@
+                rm_idx,
+                new_sz,
+                repaired,
+            );
+            repaired
+        });
+    }
+}
+
+impl<K, S, Req> ReadyCache<K, S, Req>
+where
+    K: Clone + Eq + Hash,
+    S: Service<Req>,
+    <S as Service<Req>>::Error: Into<error::Error>,
+    S::Error: Into<error::Error>,
+{
+    /// Pushes a new service onto the unready set.
+    ///
+    /// The service will be promoted to the ready set as `poll_pending` is invoked.
+    ///
+    /// Note that this does not replace a service currently in the ready set with
+    /// an equivalent key until this service becomes ready or the existing
+    /// service becomes unready.
+    pub fn push_service(&mut self, key: K, svc: S) {
+        let (tx, cancel) = oneshot::channel();
+        if let Some(c) = self.cancelations.insert(key.clone(), tx) {
+            // If there is already a pending service for this key, cancel it.
+            c.send(()).expect("cancel receiver lost");
+        }
+        self.pending.push(EvictableReady {
+            key: Some(key),
+            ready: Ready::new(svc),
+            cancel,
+        });
+    }
+
+    /// Polls unready services to readiness.
+    ///
+    /// Returns `Async::Ready` when there are no remaining unready services.
+    /// `poll_ready` should be called again after `push_service` or
+    /// `call_ready_index` are invoked.
+    pub fn poll_pending(&mut self) -> Poll<(), error::Failed<K>> {
+        loop {
+            match self.pending.poll() {
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+                Ok(Async::Ready(None)) => return Ok(Async::Ready(())),
+                Ok(Async::Ready(Some((key, svc)))) => {
+                    trace!("endpoint ready");
+                    self.cancelations.remove(&key).expect("cancel sender lost");
+                    self.ready.insert(key, svc);
+                }
+                Err(PendingError::Canceled(_)) => {
+                    debug!("endpoint canceled");
+                    // Either the cancelation has already been removed, or a
+                    // replacement unready service has already been pushed.
+                }
+                Err(PendingError::Inner(key, e)) => {
+                    self.cancelations
+                        .swap_remove(&key)
+                        .expect("missing cancelation");
+                    return Err(error::Failed(key, e.into()));
+                }
+            }
+        }
+    }
+
+    /// Configures the ready service to be used by `call`.
+    ///
+    /// This may be called instead of invoking `Service::poll_ready`
+    pub fn poll_next_ready<Q: Hash + Equivalent<K>>(
+        &mut self,
+        key: &Q,
+    ) -> Poll<(), error::Failed<K>> {
+        match self.ready.get_full_mut(key) {
+            None => Ok(Async::NotReady),
+            Some((index, _, _)) => self.poll_next_ready_index(index),
+        }
+    }
+
+    /// Consume the `next_ready_index`.
```

Comment:
A note about the implication of this on `call`, and the expected pattern of calling `poll_next_ready_index` subsequently would be good here.